### PR TITLE
make string/bytes distinguishable in the underlying compiler implementation

### PR DESCRIPTION
**Pull request by bobzhang**
This is one part of the patch required for [bucklescript](https://github.com/bloomberg/bucklescript) compiler, but I think it would be generally helpful to other backends.

The patch will be composed of such components:
1. All the lambda primtives (Pstring_, Pbytes_)
2. Runtime c function (for example, caml_blit_[bytes|string]
3. Adapt the stdlib to pick the right primitives in string/bytes module
4. Propagate the primitives in native backend, for this revision, string|bytes will be treated in the same way in native backend

It is quite a large patch, but the patch itself is not complex

### CODE REVIEW

**Comment by bobzhang** [comment]
@damiendoligez are you happy with this direction?
@chambart would you have a look at the propagation in middle_end/asmcomp? Thanks

**Comment by bobzhang** [comment]
ready for review.

two minor issues:

``` ocaml
external caml_create_string : int -> bytes
external caml_fill_string : bytes -> int -> int -> char -> unit
```

The name is a bit misleading, however, if we rename it into `caml_create_bytes`, it might break backward compatiblity, @alainfrisch what do you think?

File: bytecomp/bytegen.ml [hunk 1]
```diff
@@ -576,7 +576,7 @@ let rec comp_expr env exp sz cont =
         in
         comp_init env sz decl_size
       end
-  | Lprim((Pidentity | Popaque), [arg]) ->
+  | Lprim((Pidentity | Popaque | Pbytes_to_string | Pbytes_of_string), [arg]) ->
```
Comment (hhugo) [inline bytecomp/bytegen.ml]: Why do you treat `Pbytes_to_string` & `Pbytes_of_string` specially  ?

File: bytecomp/simplif.ml [hunk 2]
```diff
@@ -566,7 +566,7 @@ let rec emit_tail_infos is_tail lambda =
   | Lletrec (bindings, body) ->
       List.iter (fun (_, lam) -> emit_tail_infos false lam) bindings;
       emit_tail_infos is_tail body
-  | Lprim (Pidentity, [arg]) ->
+  | Lprim ((Pidentity | Pbytes_to_string | Pbytes_of_string), [arg]) ->
```
Comment (hhugo) [inline bytecomp/simplif.ml]: Why do you treat `Pbytes_to_string` & `Pbytes_of_string` specially ?

File: bytecomp/bytegen.ml (hunk 1 above)
Comment (bobzhang) [inline bytecomp/bytegen.ml]: this revision does do any change to the code gen, for both native and bytecode gen, but it is a  step to propagate it into bytecode or native backend, you can send another PR after this is merged

File: bytecomp/bytegen.ml [hunk 3]
```diff
@@ -345,11 +345,11 @@ let comp_primitive p args =
...
-  | Pstringlength -> Kccall("caml_ml_string_length", 1)
-  | Pstringrefs -> Kccall("caml_string_get", 2)
-  | Pstringsets -> Kccall("caml_string_set", 3)
-  | Pstringrefu -> Kgetstringchar
-  | Pstringsetu -> Ksetstringchar
+  | Pstringlength | Pbyteslength -> Kccall("caml_ml_string_length", 1)
+  | Pstringrefs | Pbytesrefs -> Kccall("caml_string_get", 2)
+  | Pstringsets | Pbytessets -> Kccall("caml_string_set", 3)
+  | Pstringrefu | Pbytesrefu -> Kgetstringchar
+  | Pstringsetu | Pbytessetu -> Ksetstringchar
```
Comment (hhugo) [inline bytecomp/bytegen.ml]: You should propagate this change further.

```
| Pbyteslength -> Kccall("caml_ml_byte_length", 1)
| Pbytesrefs -> Kccall("caml_byte_get", 2)
| Pbytessets -> Kccall("caml_byte_set", 3)
| Pbytesrefu -> Kgetbytechar
| Pbytessetu -> Ksetbytechar
```

File: bytecomp/simplif.ml (hunk 2 above)
Comment (bobzhang) [inline bytecomp/simplif.ml]: see my comments above, it is just to preserve the semantics for codegen **for this revision**, native or bytecode backend definitely should continue propagating such information for optimizations

File: bytecomp/bytegen.ml (hunk 1 above)
Comment (hhugo) [inline bytecomp/bytegen.ml]: Ok. Then I don't understand why you have some changes in `byterun/io.c` and `byterun/str.c`

Comment (bobzhang) [inline bytecomp/bytegen.ml]: the goal is propagating the information while **not impacting the semantics** of code generator for this revision(it already touched 25 files and needs bootstrapping), once we change the code generation, it might introduce some subtle bugs if it is not complete.
As I said, it is a good step to propagate it into bytecode/native backend once you have confidence that it is complete, but I won't do it in this revision

File: bytecomp/simplif.ml (hunk 2 above)
Comment (bobzhang) [inline bytecomp/simplif.ml]: you are right, for this simplification, we don't need it (and it will still not change the code gen) I will update

File: bytecomp/bytegen.ml (hunk 1 above)
Comment (hhugo) [inline bytecomp/bytegen.ml]: Ok. Then I don't understand why you have some changes in byterun/io.c and byterun/str.c

I understand that you don't want to change the code generation, but in that case you probably don't want to change the runtime either. that would make this revision touch 23 files only

Comment (bobzhang) [inline bytecomp/bytegen.ml]: those two primitives are used in `Pervasives.unsafe_output`, `Pervasives.unsafe_output_string`, `String.unsafe_blit`, `Bytes.unsafe_blit`
The changes are minimal

File: bytecomp/simplif.ml (hunk 2 above)
Comment (bobzhang) [inline bytecomp/simplif.ml]: @hhugo actually, if I remove this, it might introduce some performance regression here, I don't have strong opinions, I would like to hear other people's thoughts

File: bytecomp/bytegen.ml (hunk 1 above)
Comment (hhugo) [inline bytecomp/bytegen.ml]: I can see there are used.
It just feels inconsistent to only partially add primitives for `bytes`.

Comment (hhugo) [inline bytecomp/bytegen.ml]: btw, what about `external equal : t -> t -> bool = "caml_string_equal"` used in bytes.ml ?

Comment (bobzhang) [inline bytecomp/bytegen.ml]: good point, I missed it, will add it -- this is the reason I don't want to change the codegen for this revision, in case I missed something

Comment (hhugo) [inline bytecomp/bytegen.ml]: Don't modify the runtime in this feature, and you won't miss anything.
Anyway, it does not really matter.

File: otherlibs/threads/pervasives.ml [hunk 4]
```diff
@@ -195,8 +195,8 @@ external string_blit : string -> int -> bytes -> int -> int -> unit
                      = "caml_blit_string" [@@noalloc]
 external bytes_blit : bytes -> int -> bytes -> int -> int -> unit
                         = "caml_blit_string" [@@noalloc]
```
Comment (damiendoligez) [inline otherlibs/threads/pervasives.ml]: Shouldn't this be changed into `"caml_blit_bytes"` ?
Same remark for `bytes_length` and `bytes_create` a few lines above. Please review the whole file.

**Comment by damiendoligez** [comment]
I'm OK with this idea, but I'd like to hear the opinions of back-end experts (@hhugo, @mshinwell @chambart, anyone else).

One remark: I don't understand why you keep `Pstringsets` and `Pstringsetu` : the whole point of the bytes/string distinction is to get rid of them, and indeed they are not used in the standard library and not reachable by user code.

File: otherlibs/threads/pervasives.ml (hunk 4 above)
Comment (bobzhang) [inline otherlibs/threads/pervasives.ml]: good catch, I missed some functions

**Comment by bobzhang** [comment]
@damiendoligez  I will remove `Pstringsets` and `Pstringsetu`

**Comment by bobzhang** [comment]
@damiendoligez removing `Pstringsets` will break some backward compatibility, there might be some code using `%string_safe_set`, are you okay with that?

File: bytecomp/bytegen.ml (hunk 3 above)
Comment (damiendoligez) [inline bytecomp/bytegen.ml]: These should be `bytes`, not `byte`, but I'm not sure renaming the C functions is worth the trouble.

**Comment by damiendoligez** [comment]
> @damiendoligez removing `Pstringsets` will break some backward compatibility, there might be some code using `%string_safe_set`, are you okay with that?

Since it's provided as `Bytes.set` (and `String.set`) there's no excuse for using `%string_safe_set` directly, so I'm OK with breaking such code.

Or you might want to translate it into `Pbytessets` instead. In any case, a back-end that separates immutable `string` from mutables `bytes` will have to fail on `Pstringsets` (or map it to `Pbytessets` instead).

**Comment by bobzhang** [comment]
@damiendoligez removed Pstringset(s|u) per suggestions

**Comment by bobzhang** [comment]
ping @damiendoligez , is it good for merge?

File: stdlib/bytes.ml [hunk 5]
```diff
@@ -15,20 +15,20 @@
...
 (* Byte sequence operations *)
 
-external length : bytes -> int = "%string_length"
+external length : bytes -> int = "%bytes_length"
 external string_length : string -> int = "%string_length"
-external get : bytes -> int -> char = "%string_safe_get"
-external set : bytes -> int -> char -> unit = "%string_safe_set"
+external get : bytes -> int -> char = "%bytes_safe_get"
+external set : bytes -> int -> char -> unit = "%bytes_safe_set"
 external create : int -> bytes = "caml_create_string"
```
Comment (hhugo) [inline stdlib/bytes.ml]: `external create : int -> bytes = "caml_create_bytes"` ?

Comment (bobzhang) [inline stdlib/bytes.ml]: see my comments above:

``` ocaml
external caml_create_string : int -> bytes
external caml_fill_string : bytes -> int -> int -> char -> unit
```

indeed, the name is a bit misleading, either we stick to the name for better backward compatibility or we change the c primitive names or we add a duplication, I am happy with either, @hhugo @damiendoligez what do you think

Comment (hhugo) [inline stdlib/bytes.ml]: you've started partially duplicating string primitives : `caml_ml_output_bytes` `caml_blit_bytes` `caml_caml_bytes_equal`.
I would duplicate all or none.

**Comment by bobzhang** [comment]
@hhugo replaced caml_fill_string, caml_create_string per suggestions

**Comment by bobzhang** [comment]
ping @damiendoligez  @gasche ? I would say this is a safe internal change, let me know anything I can do to help it get merged, thanks!

File: byterun/str.c [hunk 6]
```diff
@@ -57,6 +60,16 @@ CAMLprim value caml_create_string(value len)
   return caml_alloc_string(size);
 }
 
+/* [len] is a value that represents a number of bytes (chars) */
+CAMLprim value caml_create_bytes(value len)
+{
+  mlsize_t size = Long_val(len);
+  if (size > Bsize_wsize (Max_wosize) - 1){
+    caml_invalid_argument("String.create");
```
Comment (alainfrisch) [inline byterun/str.c]: `"Bytes.create"`?

File: byterun/str.c [hunk 7]
```diff
@@ -229,6 +242,20 @@ CAMLprim value caml_string_equal(value s1, value s2)
   return Val_true;
 }
 
+CAMLprim value caml_bytes_equal(value s1, value s2)
```
Comment (alainfrisch) [inline byterun/str.c:264]: What's the point of duplicating the code of these primitives?  I understand it is useful to call different primitives (so that e.g. js_of_ocaml can plug different alternatives), but why duplicate the function bodies in the C runtime?

File: bytecomp/bytegen.ml [hunk 8]
```diff
@@ -345,11 +345,11 @@ let comp_primitive p args =
   | Pfloatcomp Cgt -> Kccall("caml_gt_float", 2)
   | Pfloatcomp Cle -> Kccall("caml_le_float", 2)
   | Pfloatcomp Cge -> Kccall("caml_ge_float", 2)
-  | Pstringlength -> Kccall("caml_ml_string_length", 1)
-  | Pstringrefs -> Kccall("caml_string_get", 2)
-  | Pstringsets -> Kccall("caml_string_set", 3)
-  | Pstringrefu -> Kgetstringchar
-  | Pstringsetu -> Ksetstringchar
+  | Pstringlength | Pbyteslength -> Kccall("caml_ml_string_length", 1)
```
Comment (alainfrisch) [inline bytecomp/bytegen.ml]: This PR introduces specific runtime primitives for bytes in str.c.  To be coherent, Pbyteslength should call a new `caml_ml_bytes_length` primitive.  Same for Pbytesrefs/u.

File: byterun/str.c (hunk 7 above)
Comment (bobzhang) [inline byterun/str.c:264]: you mean `return caml_string_equal(s1,s2)` internally?

File: bytecomp/bytegen.ml (hunk 8 above)
Comment (bobzhang) [inline bytecomp/bytegen.ml]: I don't want to touch bytecode/native generation in this revision, we can leave it to next revision

File: byterun/str.c (hunk 7 above)
Comment (alainfrisch) [inline byterun/str.c:264]: Yes, for instance, or use a macro to share the definition.

File: bytecomp/bytegen.ml (hunk 8 above)
Comment (alainfrisch) [inline bytecomp/bytegen.ml]: The problem is that this partial approach breaks the current coherence.  Today, the runtime system and compiler do not make any difference between strings and bytes.  With your current PR, some C primitives become specialized to bytes/strings, but not all of them.  I can see that this is useful for Bucklescript (which works at the lambda level), but it is not enough for js_of_ocaml and it leaves the system in an inconsistent state.

Formally you don't change the bytecode/native generation, but in practice, the same source code will be compiled differently because of the changes to the non-% primitives in the stdlib.  This means that js_of_ocaml will have to create aliases for its implementation of primitives, without being able to benefit from the bytes/string split.

If you want to split the proposal in two, the natural steps would rather be:
1. Do not touch the runtime system at all.  Only introduce some new internal %-primitives (forgetting the strings/bytes distinction when mapping them to C calls) and use them in the stdlib.
2. Make the runtime system aware of the distinction and duplicate all C primitives.  Adapt the remaining non-% primitives in the stdlib and the code generator to target the new primitives.

Your current proposal is between 1 and 2, just enough for Bucklescript, but I think it is worth the extra effort now to do the full 2.

Comment (bobzhang) [inline bytecomp/bytegen.ml]: I will address the comments you mentioned above, but I don't think it would be complete.
Everytime I add a new primitive I need do a bootstrapping, it takes me lots of time,  my suggestion is @hhugo could take it from here to make it complete for bytecode backend, this is a good step in anyway, right?

Comment (alainfrisch) [inline bytecomp/bytegen.ml]: I don't think that bootstrapping is required any more when primitives are added.

I'm concerned with the system be left in an inconsistent state, with the runtime system making the distinction between string and bytes primitives only for some operations.  I'm fine with the current PR if @hhugo (or someone else) is indeed willing to finish the transition.

**Comment by bobzhang** [comment]
Hi all, I addressed Alain's comments and applied Hugo's patch.  (I am not very confident to apply other patches in bytegen, but it is a good step in this direction)
Please review again, if everyone is happy, should I do the rebase against 4.04 or trunk?

**Comment by damiendoligez** [comment]
Reviewed, slightly modified, merged, and cherry-picked to 4.04.

**Comment by mshinwell** [comment]
I thought this was going to be deferred until 4.05?

**Comment by damiendoligez** [comment]
It is quite a low-risk change, useful in the short term for some users and in the long term for js_of_ocaml.

**Comment by gasche** [comment]
This PR breaks the compilation of `Flowcaml`, that uses

```
external set : ('a, 'b) t -> 'b int -> 'a char -{'b ||}-> unit 
               with 'b < 'a
= "%string_safe_set"
```

in [a module](https://github.com/pmundkur/flowcaml/blob/master/src-flowcaml/runlib/charray.ml) (the weird types are FlowCaml types, the point is the external name) and `include String` as the implementation.

opam-builder report: http://opam.ocamlpro.com/builder/html/flowcaml/flowcaml.1.07/c9f316bcefc4331699e41ea036cce35c

One may want to debate whether this is a problematic breakage or not (possibly not). I do wonder whether keeping the legacy primitive names for the `String` functions, and using the new names for the `Bytes` function would be acceptable for the backends.

Note that this PR does not seem to have a proper Changes entry (it should be marked as compatibility-breaking), the only change entry in the PR is

```
+* GPR#442, caml_fill_string, and caml_create_string is deprecated 
+  and will be removed in the future, please use caml_fill_bytes 
+  and caml_create_bytes for migration 
+  (Hongbo Zhang, review by Damien Doligez, Alain Frisch, and Hugo Heuzard)
```

which does not list this PR number and does not mention the other primitive name changes.

**Comment by bobzhang** [comment]
hi, @gasche I did discuss this issue with @damiendoligez before, shall I create a new PR to enhance Changes entry?
Btw, I thought \* means breaking change, did I miss anything?

**Comment by gasche** [comment]
Another package that is broken by this change is `rml`.

opam-builder report:
  http://opam.ocamlpro.com/builder/html/rml/rml.1.09.04/c52b6f5e8a4b821200e5ae80b2e54cc7

**Comment by gasche** [comment]
`wget` is also broken by this change

opam-builder report: http://opam.ocamlpro.com/builder/html/wget/wget.0.1.0/b51a913249637bfbe31a754953ba2135

**Comment by gasche** [comment]
@damiendoligez, @bobzhang I'm getting concerned with the number of unrelated packages hit by this issue (Batteries was also hit, of course, but that is more to be expected).

Is there anything wrong with keeping the old primitive names `caml_string_set` and others for mutable string operations, and rewriting them into the corresponding Bytes operation in the compiler code? If I understand correctly this would still be reasonable for alternative backend (there is no confusion on which types these operations should have in a `-safe-string` world), and it would remove this source of breakage.

**Comment by gasche** [comment]
If you were to both agree that this is a reasonable compromise, I could send a PR against the ocaml codebase reverting this part of the change (or @bobzhang might want to implement it directly), to be integrated in a second beta version for 4.04.

**Comment by bobzhang** [comment]
@gasche If we are moving in the direction of `-safe-string`, we are going to break the existing code any way, I would see people to fix their code sooner than later.
see the signature https://github.com/samoht/ocaml-wget/blob/master/src/stringext.mli#L19

``` ocaml
 external set : string -> int -> char -> unit = "%string_safe_set"
```

This signature does not make sense under `-safe-string`, right? How is it different that fix the code in 4.04 vs 4.05?

**Comment by gasche** [comment]
What you say is not unreasonable. I still think it's a pity that software that has been working for years is broken by a change that is not, in itself, necessary, and doesn't come with the advantages of making `-safe-string` the default, but I guess that at this point in release time in may be too late to change back anyway so maybe we should go with it.

**Comment by bobzhang** [comment]
@gasche That's my argument but I don't have strong opinions here, I can send a PR which treats `%string_safe_set` the same as `%bytes_safe_set` so it will not break other people's code. My concerns are that this actually will make the transition from `-unsafe-string` to `-safe-string` slower. I think the consensus is that in the long term that we should only support `-safe-string` . @damiendoligez what do you think?

**Comment by gasche** [comment]
One of the reasons why I'm ready to concede this specific point is that the packages that are broken by this change appear to be maintained at a very low frequency (FlowCaml hasn't changed in eight years), so it's not clear that a well-defined, well-announced migration path would actually do a big difference compared to the current breakage.

(Well-handled transitions have the property that maintainers fix the issue before it starts breaking code. Low-frequency maintainers only fix issues when they break code, so transitions don't make much of a difference.)

**Comment by damiendoligez** [comment]
@bobzhang Is is possible to simply make %string_safe_set an alias for %bytes_safe_set, so that the compilers would generate exactly the same code for both? This way, the compatibility with old code is maintained, but we can use the new name in the stdlib.

This means the legacy code won't be compatible with bucklescript or future versions of js_of_ocaml, but in a world where the back-ends need to know the difference between `string` and `bytes`, it isn't possible to mutate  a `string`, so both primitives should operate on `bytes` anyway.

**Comment by bobzhang** [comment]
>  Is is possible to simply make %string_safe_set an alias for %bytes_safe_set, so that the compilers would generate exactly the same code for both? 

@damiendoligez Yes, I will send a pull request
